internal/jobcontainers/jobcontainer.go

```diff
@@ -0,0 +1,393 @@
+package jobcontainers
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"os"
+	"os/exec"
+	"regexp"
+	"sync"
+	"syscall"
+	"time"
+
+	"github.com/Microsoft/hcsshim/internal/queue"
+	"github.com/pkg/errors"
+
+	"github.com/Microsoft/hcsshim/internal/jobobject"
+	"github.com/Microsoft/hcsshim/internal/layers"
+
+	"github.com/sirupsen/logrus"
+
+	"github.com/Microsoft/go-winio/pkg/guid"
+	"github.com/Microsoft/hcsshim/internal/cow"
+	"github.com/Microsoft/hcsshim/internal/hcs"
+	"github.com/Microsoft/hcsshim/internal/log"
+	"github.com/Microsoft/hcsshim/internal/schema1"
+	hcsschema "github.com/Microsoft/hcsshim/internal/schema2"
+	specs "github.com/opencontainers/runtime-spec/specs-go"
+	"golang.org/x/sys/windows"
+)
+
+type initProc struct {
+	initDoOnce sync.Once
+	proc       *JobProcess
+	initBlock  chan struct{}
+}
+
+// JobContainer represents a lightweight container composed from a job object.
+type JobContainer struct {
+	id             string
+	spec           *specs.Spec          // OCI spec used to create the container
+	job            *jobobject.JobObject // Object representing the job object the container owns
+	sandboxMount   string               // path to where the sandbox is mounted on the host
+	closedWaitOnce sync.Once
+	init           initProc
+	exited         chan struct{}
+	waitBlock      chan struct{}
+	waitError      error
+}
+
+var _ cow.ProcessHost = &JobContainer{}
+var _ cow.Container = &JobContainer{}
+
+func newJobContainer(id string, s *specs.Spec) *JobContainer {
+	return &JobContainer{
+		id:        id,
+		spec:      s,
+		waitBlock: make(chan struct{}),
+		init:      initProc{initBlock: make(chan struct{})},
+	}
+}
+
+// Create creates a new JobContainer from `s`.
+func Create(ctx context.Context, id string, s *specs.Spec) (_ cow.Container, err error) {
+	log.G(ctx).WithField("id", id).Debug("Creating job container")
+
+	if s == nil {
+		return nil, errors.New("Spec must be supplied")
+	}
+
+	if id == "" {
+		g, err := guid.NewV4()
+		if err != nil {
+			return nil, err
+		}
+		id = g.String()
+	}
+
+	if err := mountLayers(ctx, s); err != nil {
+		return nil, errors.Wrap(err, "failed to mount container layers")
+	}
+
+	const volumeGUIDRegex = `^\\\\\?\\(Volume)\{{0,1}[0-9a-fA-F]{8}\-[0-9a-fA-F]{4}\-[0-9a-fA-F]{4}\-[0-9a-fA-F]{4}\-[0-9a-fA-F]{12}(\}){0,1}\}(|\\)$`
```

Comment:
either my eyes are messing with me or there's an extra `\}` in the end which isn't optional. also, is `(|\\)` same as `\\?` (or maybe `(\\)?` for more clarity).
maybe it's worth to split this into two:
```
GUIDRegex = `[0-9a-fA-F]{8}...`
volumeGUIDRegex=fmt.Sprintf(`^\\\\\?\\(Volume)(\{%s\}|%s)\\?`, GUIDRegex, GUIDRegex)
```